From ab90dcd9c5c33d3c26f2553f63dcd1837ce49d79 Mon Sep 17 00:00:00 2001
From: Roman Stratiienko <roman.stratiienko@globallogic.com>
Date: Wed, 28 Nov 2018 14:23:57 +0200
Subject: [PATCH] mali: support building against 4.20

From 4.20 kernel vm_insert_pfn was dropped
This makes wrapper to preserve compatibility

Also get_monotonic_boottime(...) was removed from 4.20
Replace it with ktime_get_boot_ns()

Signed-off-by: Roman Stratiienko <roman.stratiienko@globallogic.com>
---
 src/devicedrv/mali/linux/mali_kernel_linux.c | 11 +++++++++++
 src/devicedrv/mali/linux/mali_kernel_linux.h | 10 ++++++++++
 src/devicedrv/mali/linux/mali_osk_time.c     |  4 ++++
 3 files changed, 25 insertions(+)

diff --git a/src/devicedrv/mali/linux/mali_kernel_linux.c b/src/devicedrv/mali/linux/mali_kernel_linux.c
index 34cb3d7..6ed89a6 100755
--- a/src/devicedrv/mali/linux/mali_kernel_linux.c
+++ b/src/devicedrv/mali/linux/mali_kernel_linux.c
@@ -1125,6 +1125,17 @@
 	return err;
 }
 
+/*
+ * From 4.20.0 kernel vm_insert_pfn was dropped
+ * Make wrapper to preserve compatibility
+ */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 20, 0)
+int vm_insert_pfn(struct vm_area_struct *vma, unsigned long addr,
+		  unsigned long pfn)
+{
+	return vm_fault_to_errno(vmf_insert_pfn(vma, addr, pfn), 0xffff);
+}
+#endif
 
 module_init(mali_module_init);
 module_exit(mali_module_exit);
diff --git a/src/devicedrv/mali/linux/mali_kernel_linux.h b/src/devicedrv/mali/linux/mali_kernel_linux.h
index 8548c9f..f858a5c 100755
--- a/src/devicedrv/mali/linux/mali_kernel_linux.h
+++ b/src/devicedrv/mali/linux/mali_kernel_linux.h
@@ -16,6 +16,7 @@
 #endif
 
 #include <linux/cdev.h>     /* character device definitions */
+#include <linux/mm.h>
 #include <linux/idr.h>
 #include <linux/rbtree.h>
 #include "mali_kernel_license.h"
@@ -33,6 +34,15 @@
 #define __GFP_REPEAT __GFP_RETRY_MAYFAIL
 #endif
 
+/*
+ * From 4.20.0 kernel vm_insert_pfn was dropped
+ * Make wrapper to preserve compatibility
+ */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 20, 0)
+extern int vm_insert_pfn(struct vm_area_struct *vma, unsigned long addr,
+			 unsigned long pfn);
+#endif
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/src/devicedrv/mali/linux/mali_osk_time.c b/src/devicedrv/mali/linux/mali_osk_time.c
index 4deaa10..a14cb7d 100755
--- a/src/devicedrv/mali/linux/mali_osk_time.c
+++ b/src/devicedrv/mali/linux/mali_osk_time.c
@@ -53,7 +53,11 @@ u64 _mali_osk_time_get_ns(void)
 
 u64 _mali_osk_boot_time_get_ns(void)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 20, 0)
+	return ktime_get_boot_ns();
+#else
 	struct timespec tsval;
 	get_monotonic_boottime(&tsval);
 	return (u64)timespec_to_ns(&tsval);
+#endif
 }
-- 
2.17.1

